File: updater.py
import tkinter as tk
from tkinter import ttk, messagebox
import os
import threading
import shutil
import traceback
from datetime import datetime

import config
import utils
from installer import copy_instance_files

def backup_world(dst_path, update_status, log_func):
    update_status("Backing up World...", 0)
    world_name = "world"
    props_path = os.path.join(dst_path, "server.properties")
    
    if os.path.exists(props_path):
        try:
            with open(props_path, "r") as f:
                for line in f:
                    if line.strip().startswith("level-name="):
                        world_name = line.strip().split("=", 1)[1]
                        break
        except Exception as e:
            log_func(f"Failed to read server.properties: {e}", error=True)
            
    world_dir = os.path.join(dst_path, world_name)
    if not os.path.exists(world_dir):
        log_func(f"World folder '{world_name}' not found. Skipping backup.")
        return
        
    backups_dir = os.path.join(dst_path, "backups")
    if not os.path.exists(backups_dir):
        os.makedirs(backups_dir)
        
    timestamp = datetime.now().strftime("%Y-%m-%d_%H-%M-%S")
    backup_filename = f"{world_name}_{timestamp}"
    backup_path = os.path.join(backups_dir, backup_filename)
    
    try:
        log_func(f"Backing up world '{world_name}' to backups/{backup_filename}.zip...")
        shutil.make_archive(backup_path, 'zip', world_dir)
        log_func("World backup complete.")
    except Exception as e:
        log_func(f"World backup failed: {e}", error=True)

def run_update(src_path, dst_path, update_status, log_func, on_success, on_failure):
    try:
        backup_world(dst_path, update_status, log_func)
        copy_instance_files(src_path, dst_path, update_status, log_func)
        update_status("Complete!", 100)
        on_success()
    except Exception as e:
        trace = traceback.format_exc()
        log_func(f"CRASH: {trace}", error=True)
        on_failure(str(e))

class UpdateApp:
    def __init__(self, root, src_path, dst_path):
        self.root = root
        self.root.title("Update Server - Farming Experience Server Installer")
        
        try:
            self.root.iconbitmap(utils.resource_path("app.ico"))
        except Exception:
            pass
        
        utils.center_window(root, 400, 300)
        self.root.resizable(False, False)
        
        self.log_buffer = []
        self.src_path = src_path
        self.dst_path = dst_path
        
        style = ttk.Style()
        try:
            style.theme_use('vista')
        except Exception:
            style.theme_use('clam')
        style.configure("TProgressbar", thickness=15)

        header_frame = tk.Frame(root, bg="#2E7D32")
        header_frame.pack(fill="x", side="top")
        
        lbl_title = tk.Label(header_frame, text="Farming Experience", font=("Segoe UI", 18, "bold"), bg="#2E7D32", fg="white")
        lbl_title.pack(pady=(15, 2))
        
        lbl_subtitle = tk.Label(header_frame, text="Update Server", font=("Segoe UI", 10), bg="#2E7D32", fg="#E8F5E9")
        lbl_subtitle.pack(pady=(0, 15))

        self.content_area = tk.Frame(root, padx=20, pady=20)
        self.content_area.pack(fill="both", expand=True)

        if not src_path or not dst_path:
            tk.Label(self.content_area, text="Invalid parameters passed to updater.", fg="red").pack()
            return
            
        self.lbl_ready = tk.Label(self.content_area, text="Ready to update your server mods and configs?", font=("Segoe UI", 10))
        self.lbl_ready.pack(pady=(0, 20))
        
        self.btn_update = tk.Button(self.content_area, text="UPDATE NOW", **config.PRIMARY_BTN_STYLE, command=self.start_update)
        self.btn_update.pack()

    def log(self, message, error=False):
        utils.log_message(self.log_buffer, message, error)
        
    def update_status(self, text, percent=None, sub_text=""):
        self.lbl_status.config(text=text)
        if percent is not None:
            self.progress_var.set(percent)
        self.root.update_idletasks()

    def start_update(self):
        self.btn_update.pack_forget()
        self.lbl_ready.pack_forget()
        
        self.progress_container = tk.Frame(self.content_area)
        self.progress_container.pack(fill="x", pady=10)

        self.lbl_status = tk.Label(self.progress_container, text="Preparing...", font=("Segoe UI", 10), fg="#333")
        self.lbl_status.pack(fill="x", pady=(0, 5))
        
        self.progress_var = tk.DoubleVar()
        self.progress_bar = ttk.Progressbar(self.progress_container, variable=self.progress_var, maximum=100)
        self.progress_bar.pack(fill="x")
        
        threading.Thread(
            target=run_update,
            args=(
                self.src_path,
                self.dst_path,
                self.update_status,
                self.log,
                self.on_update_success,
                self.on_update_failure
            ),
            daemon=True
        ).start()

    def on_update_success(self):
        self.root.after(0, self._show_success_ui)
        utils.flush_logs(os.path.join(self.dst_path, "serverinstaller"), self.log_buffer, prefix="update")

    def on_update_failure(self, error_message):
        self.root.after(0, lambda: messagebox.showerror("Error", f"Update Failed.\nCheck the logs folder.\n\n{error_message}"))
        self.root.after(0, self.root.destroy)
        utils.flush_logs(os.path.join(self.dst_path, "serverinstaller"), self.log_buffer, prefix="update")

    def _show_success_ui(self):
        self.progress_container.pack_forget()
        tk.Label(self.content_area, text="Update Complete!", font=("Segoe UI", 14, "bold"), fg="#2E7D32").pack(pady=(10, 20))
        btn_close = tk.Button(self.content_area, text="CLOSE", **config.PRIMARY_BTN_STYLE, command=self.root.destroy)
        btn_close.pack()

File: src/main/java/com/farmingexperience/installer/Updater.java
package com.farmingexperience.installer;

import javax.swing.*;
import java.awt.*;
import java.io.File;
import java.io.IOException;
import java.io.OutputStream;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;
import java.util.List;
import java.util.zip.ZipEntry;
import java.util.zip.ZipOutputStream;

public final class Updater {
    private Updater() {}

    public static void backupWorld(String dstPath, Installer.StatusFunc updateStatus, Utils.LogFunc logFunc) {
        updateStatus.update("Backing up World...", 0.0);
        String worldName = "world";
        Path propsPath = Path.of(dstPath, "server.properties");

        if (Files.exists(propsPath)) {
            try {
                // this just in case the world name isn't "world"
                for (String line : Files.readAllLines(propsPath, StandardCharsets.UTF_8)) {
                    if (line.strip().startsWith("level-name=")) {
                        worldName = line.strip().split("=", 2)[1];
                        break;
                    }
                }
            } catch (Exception e) {
                logFunc.log("Failed to read server.properties: " + e.getMessage(), true);
            }
        }

        Path worldDir = Path.of(dstPath, worldName);
        if (!Files.exists(worldDir)) {
            logFunc.log("World folder '" + worldName + "' not found. Skipping backup.");
            return;
        }

        Path backupsDir = Path.of(dstPath, "backups");
        try {
            Files.createDirectories(backupsDir);
        } catch (IOException e) {
            logFunc.log("Failed to create backups folder: " + e.getMessage(), true);
            return;
        }

        String timestamp = LocalDateTime.now().format(DateTimeFormatter.ofPattern("yyyy-MM-dd_HH-mm-ss"));
        String backupFilename = worldName + "_" + timestamp;
        Path backupZip = backupsDir.resolve(backupFilename + ".zip");

        try {
            logFunc.log("Backing up world '" + worldName + "' to backups/" + backupFilename + ".zip...");
            zipDirectory(worldDir, backupZip);
            logFunc.log("World backup complete.");
        } catch (Exception e) {
            logFunc.log("World backup failed: " + e.getMessage(), true);
        }
    }

    private static void zipDirectory(Path sourceDir, Path zipPath) throws IOException {
        try (OutputStream fos = Files.newOutputStream(zipPath);
             ZipOutputStream zos = new ZipOutputStream(fos)) {
            addFolderToZip(sourceDir, sourceDir, zos);
        }
    }

    private static void addFolderToZip(Path rootDir, Path currentDir, ZipOutputStream zos) throws IOException {
        File[] children = currentDir.toFile().listFiles();
        if (children == null) {
            return;
        }

        for (File child : children) {
            Path childPath = child.toPath();
            if (child.isDirectory()) {
                addFolderToZip(rootDir, childPath, zos);
            } else {
                String entryName = rootDir.relativize(childPath).toString().replace('\\', '/');
                zos.putNextEntry(new ZipEntry(entryName));
                Files.copy(childPath, zos);
                zos.closeEntry();
            }
        }
    }

    public static void runUpdate(String srcPath, String dstPath, Installer.StatusFunc updateStatus,
                                  Utils.LogFunc logFunc, Runnable onSuccess, Installer.FailureCallback onFailure) {
        try {
            backupWorld(dstPath, updateStatus, logFunc);
            Installer.copyInstanceFiles(srcPath, dstPath, updateStatus, logFunc);
            updateStatus.update("Complete!", 100.0);
            onSuccess.run();
        } catch (Exception e) {
            logFunc.log("CRASH: " + Utils.stackTraceToString(e), true);
            onFailure.onFailure(e.getMessage());
        }
    }

    public static class UpdateApp {
        private final JFrame frame;
        private final List<String> logBuffer = new ArrayList<>();
        private final String srcPath;
        private final String dstPath;

        private JPanel contentArea;
        private JLabel lblReady;
        private JButton btnUpdate;
        private JPanel progressContainer;
        private JLabel lblStatus;
        private JProgressBar progressBar;

        public UpdateApp(JFrame frame, String srcPath, String dstPath) {
            this.frame = frame;
            this.srcPath = srcPath;
            this.dstPath = dstPath;

            frame.setTitle("Update Server - " + Config.APP_NAME + " Server Installer");
            Image icon = Utils.loadAppIcon();
            if (icon != null) {
                frame.setIconImage(icon);
            }

            int w = 460, h = 340;
            Utils.centerWindow(frame, w, h);
            frame.setResizable(false);
            frame.setLayout(new BorderLayout());

            log("Updater started.");

            JPanel headerFrame = new JPanel();
            headerFrame.setBackground(Config.HEADER_BG);
            headerFrame.setLayout(new BoxLayout(headerFrame, BoxLayout.Y_AXIS));
            headerFrame.setBorder(BorderFactory.createEmptyBorder(20, 0, 20, 0));

            JLabel lblTitle = UI.label(Config.APP_NAME, Config.FONT_TITLE, Color.WHITE);
            lblTitle.setAlignmentX(Component.CENTER_ALIGNMENT);

            JLabel lblSubtitle = UI.label("Update Server", Config.FONT_SUBTITLE, Config.HEADER_SUBTITLE_FG);
            lblSubtitle.setAlignmentX(Component.CENTER_ALIGNMENT);
            lblSubtitle.setBorder(BorderFactory.createEmptyBorder(4, 0, 0, 0));

            headerFrame.add(lblTitle);
            headerFrame.add(lblSubtitle);
            frame.add(headerFrame, BorderLayout.NORTH);

            contentArea = new JPanel();
            contentArea.setBorder(BorderFactory.createEmptyBorder(20, 20, 20, 20));
            contentArea.setLayout(new BoxLayout(contentArea, BoxLayout.Y_AXIS));
            frame.add(contentArea, BorderLayout.CENTER);

            if (srcPath == null || srcPath.isEmpty() || dstPath == null || dstPath.isEmpty()) {
                JLabel lblError = UI.label("Invalid parameters passed to updater.", Config.FONT_BODY, Config.ERROR_FG);
                contentArea.add(lblError);
                frame.setVisible(true);
                return;
            }

            lblReady = UI.label("Ready to update your server mods and configs?", Config.FONT_BODY, Config.TEXT_DARK);
            lblReady.setAlignmentX(Component.CENTER_ALIGNMENT);
            lblReady.setBorder(BorderFactory.createEmptyBorder(0, 0, 24, 0));
            contentArea.add(lblReady);

            btnUpdate = UI.primaryButton("UPDATE NOW");
            btnUpdate.setAlignmentX(Component.CENTER_ALIGNMENT);
            btnUpdate.addActionListener(e -> startUpdate());
            contentArea.add(btnUpdate);

            frame.setVisible(true);
        }

        private void log(String message) {
            Utils.logMessage(logBuffer, message, false);
        }

        private void updateStatus(String text, Double percent, String subText) {
            SwingUtilities.invokeLater(() -> {
                lblStatus.setText(text);
                if (percent != null) {
                    progressBar.setValue(percent.intValue());
                }
            });
        }

        private void startUpdate() {
            contentArea.remove(btnUpdate);
            contentArea.remove(lblReady);

            progressContainer = new JPanel();
            progressContainer.setLayout(new BoxLayout(progressContainer, BoxLayout.Y_AXIS));
            progressContainer.setBorder(BorderFactory.createEmptyBorder(10, 0, 10, 0));

            lblStatus = UI.label("Preparing...", Config.FONT_BODY, Config.TEXT_DARK);
            lblStatus.setAlignmentX(Component.LEFT_ALIGNMENT);
            lblStatus.setBorder(BorderFactory.createEmptyBorder(0, 0, 6, 0));

            progressBar = new JProgressBar(0, 100);
            progressBar.setAlignmentX(Component.LEFT_ALIGNMENT);
            progressBar.setPreferredSize(new Dimension(Integer.MAX_VALUE, 20));
            progressBar.setMaximumSize(new Dimension(Integer.MAX_VALUE, 20));

            progressContainer.add(lblStatus);
            progressContainer.add(progressBar);
            contentArea.add(progressContainer);
            contentArea.revalidate();
            contentArea.repaint();

            Thread thread = new Thread(() -> runUpdate(
                    srcPath, dstPath,
                    this::updateStatus,
                    (msg, err) -> Utils.logMessage(logBuffer, msg, err),
                    this::onUpdateSuccess,
                    this::onUpdateFailure
            ));
            thread.setDaemon(true);
            thread.start();
        }

        private void onUpdateSuccess() {
            SwingUtilities.invokeLater(this::showSuccessUi);
            Utils.flushLogs(Path.of(dstPath, "serverinstaller").toString(), logBuffer, "update");
        }

        private void onUpdateFailure(String errorMessage) {
            SwingUtilities.invokeLater(() -> {
                JOptionPane.showMessageDialog(frame,
                        "Update Failed.\nCheck the logs folder.\n\n" + errorMessage,
                        "Error", JOptionPane.ERROR_MESSAGE);
                frame.dispose();
            });
            Utils.flushLogs(Path.of(dstPath, "serverinstaller").toString(), logBuffer, "update");
        }

        private void showSuccessUi() {
            contentArea.remove(progressContainer);

            JLabel lblDone = UI.label("Update Complete!", Config.FONT_HEADING, Config.HEADER_BG);
            lblDone.setAlignmentX(Component.CENTER_ALIGNMENT);
            lblDone.setBorder(BorderFactory.createEmptyBorder(12, 0, 24, 0));
            contentArea.add(lblDone);

            JButton btnClose = UI.primaryButton("CLOSE");
            btnClose.setAlignmentX(Component.CENTER_ALIGNMENT);
            btnClose.addActionListener(e -> frame.dispose());
            contentArea.add(btnClose);

            contentArea.revalidate();
            contentArea.repaint();
        }
    }
}
